ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+              + segmentCapacityInBytes + "]");
     }
-    this.capacityInBytes = capacityInBytes;
-    fileChannel = Utils.openChannel(file, true);
-    logger.trace("Log : {} file size on start {} ", dataDir, fileChannel.size());
-    // A log's write offset will always be set to the start of the log.
-    // External components is responsible for setting it the right value
-    currentWriteOffset = new AtomicLong(0);
+    this.dataDir = dataDir;
+    this.capacityInBytes = totalCapacityInBytes;
     this.metrics = metrics;
+    remainingUnallocatedCapacity = new AtomicLong(totalCapacityInBytes);
+    long numSegments = (totalCapacityInBytes - 1) / segmentCapacityInBytes + 1;
+    loadSegments(segmentCapacity, numSegments);
   }
 
-  StoreMessageReadSet getView(List<BlobReadOptions> readOptions)
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Appends the given {@code buffer} to the active log segment. The {@code buffer} will be written to a single log
+   * segment i.e. its data will not exist across segments.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if the {@code buffer.remaining()} is greater than a single segment's size.
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if there was an I/O error while writing.
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
       throws IOException {
-    return new StoreMessageReadSet(file, fileChannel, readOptions, currentWriteOffset.get());
+    rollOverIfRequired(buffer.remaining());
+    return activeSegment.appendFrom(buffer);
   }
 
-  public long sizeInBytes()
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Appends the given data to the active log segment. The data will be written to a single log segment i.e. the data
+   * will not exist across segments.
+   * @param channel The channel from which data needs to be written from
+   * @param size The amount of data in bytes to be written from the channel
+   * @throws IllegalArgumentException if the {@code size} is greater than a single segment's size.
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if there was an I/O error while writing.
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size)
       throws IOException {
-    return fileChannel.size();
+    rollOverIfRequired(size);
+    activeSegment.appendFrom(channel, size);
   }
 
-  public void setLogEndOffset(long endOffset)
-      throws IOException {
-    long fileSize = fileChannel.size();
-    if (endOffset < 0 || endOffset > fileSize) {
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " endOffset " + endOffset +
-          " outside the file size " + fileSize);
+  /**
+   * Sets the active segment in the log.
+   * @param name the name of the log segment that is to be marked active.
+   */
```

Comment:
The index will call this function to set the current active segment. That situation will arise when a segment has been allocated for an operation but a crash occurs just before the operation can actually proceed. As discussed, I will work on maintaining the invariant that the active segment is always the last segment.